### DISCUSSION TITLE
How to do mutation using query-core?

### DISCUSSION THREAD

**Discussion by thinkkevin** [TanStack/query#4807] [Q&A]
I wonder what the recommended approach is to do mutation only using `@tanstack/query-core`? I couldn't find any documentation regarding that. 

Please advise, thanks!

**Comment by TkDodo** [comment]
```
queryClient
    .getMutationCache()
    .build(queryClient, options)
    .execute()
```

but the api will change slightly

**Reply by TkDodo** [reply]
actually, I'm not sure what's the point in executing a mutation like that directly. It's not better than doing `fetch` without it. What you'd want is a `MutationObserver` that can track the state and lifecycle (loading / success) of your mutation. That is what `useMutation` does, but you can also create an observer manually. depends on what the use case is.

**Reply by thinkkevin** [reply]
@TkDodo Thanks for your reply.
I am considering building an API SDK that uses a QueryClient. The client of the SDK might use it outside of React or another UI stack. 

btw. Is API change (you mentioned) planned in v5 or v4?

**Reply by TkDodo** [reply]
v5

**Reply by blasdfaa** [reply]
> ```
> queryClient
>     .getMutationCache()
>     .build(queryClient, options)
>     .execute()
> ```
> 
> but the api will change slightly

Should this also be done in `queryClient` using `@tanstack/react-query`? I need to make a post request in `action` in react-router6, but I can't figure out how to call a mutation without a hook.

In the official example sandbox, all requests are fakes

**Reply by TkDodo** [reply]
you need to have a queryClient one way or the other, and you can create it with `new QueryClient()`. That class can be imported form either `@tanstack/query-core` or from `@tanstack/react-query`. All the framework adapters _include_ everything from the core.

**Reply by TkDodo** [reply]
> I need to make a post request in action in react-router6, but I can't figure out how to call a mutation without a hook.

Not really a need to combine actions with mutations. You could just as well fire the request _without_ a mutation I believe

**Reply by blasdfaa** [reply]
> > I need to make a post request in action in react-router6, but I can't figure out how to call a mutation without a hook.
> 
> Not really a need to combine actions with mutations. You could just as well fire the request _without_ a mutation I believe

It might even be easier to execute request without mutation. But I was wildly wondering why queryClient doesn't have such an opportunity? I didn't find such an opportunity on this page - https://tanstack.com/query/v4/docs/react/reference/QueryClient

**Reply by TkDodo** [reply]
we had one called `executeMutation` but it was unused and undocumented and no one needed it :)

https://tanstack.com/query/v4/docs/react/guides/migrating-to-react-query-4#removed-undocumented-methods-from-the-queryclient-query-and-mutation

**Reply by blasdfaa** [reply]
> we had one called `executeMutation` but it was unused and undocumented and no one needed it :)
> 
> https://tanstack.com/query/v4/docs/react/guides/migrating-to-react-query-4#removed-undocumented-methods-from-the-queryclient-query-and-mutation

Ohhh. I planned to use one approach to writing queries with react-query. Then you will have to break it in what places 

Thank you for your help!
